content/post/2022-09-01-r-dependency/index.Rmd

```diff
@@ -0,0 +1,284 @@
+As a package developer, [you may want to use one of these newly added features](https://github.com/yihui/knitr/issues/2100).
+
+In this situation, you should inform users ([as well as automated checks from CRAN](https://www.mail-archive.com/r-package-devel@r-project.org/msg06331.html)) that your package only works for R versions more recent than a given number [^1].
+
+[^1]: Note that there is no mechanism to make your package compatible only with older R versions, and not with the more recent ones.
+    Packages are supposed to work with the latest R versions.
+
+To do so, you should add the required version number to your `DESCRIPTION` file [^2]:
+
+[^2]: In theory, it is not strictly required to use `>=`.
+    You could use a strict inequality (`>`) but as we will see later, this is a very uncommon option so we recommend you use the *de facto* community standard and stick to `>=`.
+
+``` yaml
+  Depends:
+    R (>= 3.5.0)
+```
+
+# How to choose on which version you should depend?
+
+There are different strategies to choose on which R version your package should depend:
+
+## Conservative approach
+
+Some projects prefer to limit the minimum R version by design, rather than by necessity.
+This means that their packages *might* work with older R versions, but because they don't or can't test it, they'd rather not take the risk and limit themselves to versions for which they are sure the package is working:
+
+-   this used to be the policy of usethis before 2017 (and therefore, of all packages built with usethis at that time).
+    [In the past, usethis added by default a dependency to the R version used by the developer at the time they created the package](https://github.com/r-lib/usethis/commit/7937594cb4a6adc9f1783839c4ccdd2cdcffaaae).
+
+-   this is the strategy used by the tidyverse, which explicitly decided to guarantee compatibility with the 5 latest R minor releases, but no further.
+    With the current R release cycle, this corresponds to compatibility with R versions up to 5 years old.
+
+## 'Wide net' approach
+
+On the opposite, other projects consider that packages are by default compatible with all R versions, until they explicitly add a feature associated with a new R version, or until tests prove it otherwise.
+This is the new policy of usethis (and therefore, of all packages built this usethis).
+By default, new packages don't have any constraints on the R version.
+It is the responsibility of the developer to add a minimum required version if necessary.
+
+## Transitive approach
+
+Another approach is to look at your dependency.
+If indirectly, via one of your recursive dependencies, you already depend on a recent R version, there is no point in going the extra mile to keep working with older versions.
+So, a strategy could be to compute your transitive minimum R version with the following function and decide that you can use base R features up to this version:
+
+```{r}
+find_transitive_minR <- function(package) {
```

Comment:
could the function also return which of the recursive dependency created the constraint? no problem if too complicated

Reply:
It's not too difficult in terms of code but what format do you have in mind for the output? I tend to avoid list when possible because I believe they are confusing for new R users (they were for me, at least). I don't think a message is great either because users might want to be able to catch the relevant package name(s) programmatically :thinking:

Reply:
then we can forget this :-) or a data.frame? not crucial in any case